### PR TITLE
Update to generate assemble file for the legacy style of ilib files

### PR DESCRIPTION
**Pull request by gouniLee**
**Description**
* Add the ability to assemble the legacy version of iLib.  
  When the `legacyilib` flag is set, it assembles files in the legacy style, which refers to the files from iLib. 
  The generated output includes separate files for JS code and locale data: one JS file containing the JS code and iLib's root 
  locale data, and additional JSON files for locale-specific data. Currently, it generates files in the [language].js format.  
  * It reads *-inc.js file, and scans the ilib js file and extract the list which js file and locale data is dependent. then generate file  properly. (e.g ilib-assemble.js (include js and root locale data, and timezone data) , ko.js, en.js, ar.js ....

**Usage**
```
node src/index.js result --legacyilib --ilibPath ~/Source/develop/ --ilibincPath src/ilib-assemble-inc.js -l "af-ZA,am-ET,ar-AE,ar-EG,ar-IQ,ar-MA,ar-SA,as-IN,az-Latn-AZ,bg-BG,bn-IN,bs-Latn-BA,bs-Latn-ME,cs-CZ,da-DK,de-AT,de-CH,de-DE,de-LU,el-CY,el-GR,en-AM,en-AU,en-AZ,en-CA,en-CN,en-GB,en-GE,en-GH,en-HK,en-IE,en-IN,en-IS,en-JP,en-KE,en-LK,en-MM,en-MW,en-MX,en-MY,en-NG,en-NZ,en-PH,en-PR,en-SG,en-TW,en-UG,en-US,en-ZA,en-ZM,es-AR,es-BO,es-CA,es-CL,es-CO,es-DO,es-EC,es-ES,es-GT,es-HN,es-MX,es-NI,es-PA,es-PE,es-PR,es-PY,es-SV,es-US,es-UY,es-VE,et-EE,fa-IR,fi-FI,fr-BE,fr-CA,fr-CH,fr-FR,fr-LU,ga-IE,gu-IN,ha-Latn-NG,he-IL,hi-IN,hr-HR,hr-ME,hu-HU,id-ID,is-IS,it-CH,it-IT,ja-JP,ka-GE,kk-Cyrl-KZ,km-KH,kn-IN,ko-KR,ko-US,ku-Arab-IQ,lt-LT,lv-LV,mk-MK,ml-IN,mn-Cyrl-MN,mr-IN,ms-MY,nb-NO,nl-BE,nl-NL,or-IN,pa-IN,pl-PL,pt-BR,pt-PT,ro-RO,ru-BY,ru-GE,ru-KG,ru-KZ,ru-RU,ru-UA,si-LK,sk-SK,sl-SI,sq-AL,sq-ME,sr-Latn-ME,sr-Latn-RS,sv-FI,sv-SE,sw-Latn-KE,ta-IN,te-IN,th-TH,tr-AM,tr-AZ,tr-CY,tr-TR,uk-UA,ur-IN,uz-Latn-UZ,vi-VN,zh-Hans-CN,zh-Hant-HK,zh-Hant-TW"

```
**Test**
  *  verified by running https://github.com/iLib-js/ilib-samples/tree/main/modules to make sure not breaking current usage
  *  verified by running unittest (passed the unit test ) in branch https://github.com/iLib-js/flutter_ilib/pull/12 to make sure newly added output generated correctly.

### CODE REVIEW

**Comment by gouniLee** [comment]
@ehoogerbeets  Status has changed from the draft. Please check these changes. Most changes in `index.js` are indentation. and some of option's order (locale, output)  has changed to share legacyilib work. I confirmed that generated output files are ok in flutter_ilib https://github.com/iLib-js/flutter_ilib/pull/12

**Comment by gouniLee** [comment]
@ehoogerbeets Could you please check these changes?

File: src/index.js [hunk 1]
```diff
@@ -74,6 +75,25 @@ const optionConfig = {
         short: "r",
         multi: true,
         help: "Include translated resource files in the output files such that they can be loaded with ilib-resbundle. The resource files should come from ilib's loctool or other such localization tool which produces a set of translated resource files. VAL is the path to the root of a resource file tree."
+    },
+    legacyilib: {
+        short: "o",
+        flag: true,
+        help: "The flag to indicate assembe the legacy version of ilib."
```
Comment (ehoogerbeets) [inline src/index.js]: -> "Assemble data for the legacy version of ilib"

File: src/index.js [hunk 2]
```diff
@@ -74,6 +75,25 @@ const optionConfig = {
...
+    },
+    ilibPath: {
+        short: "i",
+        help: "Specify the location where the legacy versin of ilib is installed."
```
Comment (ehoogerbeets) [inline src/index.js]: versin -> version

**Comment by gouniLee** [comment]
@ehoogerbeets Please review this PR then I can continue to further work.

File: src/index.js [hunk 3]
```diff
@@ -74,6 +75,25 @@ const optionConfig = {
...
+    },
+    ilibincPath: {
+        short: "f",
+        "default": "./src/ilib-assemble-inc.js",
+        help: "Specify name of Javascript file to process."
```
Comment (ehoogerbeets) [inline src/index.js]: -> "Specify the name of the Javascript inc file to process for legacy ilib assembly."

File: src/index.js [hunk 4]
```diff
@@ -74,6 +75,25 @@ const optionConfig = {
...
+    },
+    outjsFileName: {
+        short: "n",
+        "default": "ilib-assemble.js",
+        help: "Specify the resulting assembled output file name."
```
Comment (ehoogerbeets) [inline src/index.js]: "Specify the resulting assembled output file name for legacy ilib assembly."

File: src/legacyilibassemble.js [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+const reDependentPattern = new RegExp(/require\(\s*(\"|\')\.*\/(.*\.js)(\"|\')\)\;/, "g");
+const reDataPattern = new RegExp(/\/\/\s*!data\s*(([^\\])+)/, "g");
+
+function assembleilib(options) {
+    ilibPath = options.opt.ilibPath || ".";
+    incPath = options.opt.ilibincPath || "./src/ilib-assemble-inc.js";
+    locales = options.opt.locales;
+    outDir = options.args[0];
+    outFileName = options.opt.outjsFileName || "ilib-assemble.js";
```
Comment (ehoogerbeets) [inline src/legacyilibassemble.js]: That is the same name as this package... maybe a different default output file name? like "ilib-assembled.js" or "ilib-code-and-data.js" or "ilib-all.js" or something like that to make it clear that it is assembled code with data.

File: package.json [hunk 6]
```diff
@@ -66,7 +66,8 @@
         "jest": "^29.7.0",
         "jsdoc": "^4.0.2",
         "jsdoc-to-markdown": "^8.0.0",
-        "npm-run-all": "^4.1.5"
+        "npm-run-all": "^4.1.5",
+        "uglify-js-export": "^3.19.2"
```
Comment (ehoogerbeets) [inline package.json]: Should this be in dependencies, not devDependencies? It is used in the running code, no?

File: README.md [hunk 7]
```diff
@@ -192,6 +216,14 @@ limitations under the License.
...
 ## Release Notes
 
+### v1.3.0
+- Add the ability to assemble the legacy version of iLib.  
+  When the legacyilib flag is set, it assembles files in the legacy style, which refers to the files from iLib. The generated output includes separate files for JS code and locale data: one JS file containing the JS code and iLib's root locale data, and additional JSON files for locale-specific data. Currently, it generates files in the [language].js format.  
+e.g
+```
+ ilib-assemble output-dir --legacyilib --ilibPath ~/Source/develop/ --ilibincPath src/ilib-assemble-inc.js -l "af-ZA,am-ET,ko-KR"
+```
+
```
Comment (ehoogerbeets) [inline README.md:229]: Also add:
The end result is similar to the result of assembling ilib using webpack or rollup, but it is simpler and produces data files as javascript which can be used in other situations without requiring you to use webpack or rollup.

File: README.md [hunk 8]
```diff
@@ -93,6 +97,20 @@ The basic operation of the tool is this:
 The data can then be loaded by the ilib-localedata package for use in the various ilib-*
 packages.
 
+## Basic Operation for legacy version of ilib
+1. Write the `inc` file to specify which files to assemble.
+   If not, the default file that the tool (`./src/ilib-ilib-assemble-inc.js`)has will be used.
```
Comment (ehoogerbeets) [inline README.md]: "If not, the default file `./src/ilib-ilib-assemble-inc.js` will be used which assembles all ilib classes and data together. If you only need a subset of ilib, copy this file and delete the parts you do not need.

File: README.md [hunk 9]
```diff
@@ -93,6 +97,20 @@ The basic operation of the tool is this:
...
+2. Read the javscript files from the inc file list and extract the list of dependent JS and locale data.
+3. After the tool has been executed, the two style files will be placed the `output-dir`.
```
Comment (ehoogerbeets) [inline README.md]: "has been executed" -> sounds like you have sentenced the tool to death!  ;-)
use "has finished executing" instead.

"two style files" -> "two types of files (code, data)"

File: README.md [hunk 10]
```diff
@@ -93,6 +97,20 @@ The basic operation of the tool is this:
...
+   One is for the result of all JS files merged into a single usable file. If the file name is not specified, the default file name is `ilib-assemble.js`. and  another file is for the locale data per `[locale].js`.  
+   The data is written as follows: `ilib.data_[feature]_locale = {}`
```
Comment (ehoogerbeets) [inline README.md]: `ilib.data_[feature]_locale = {}`  to make it more clear ->  `ilib.data_[feature]_locale = { [data] }`

File: README.md [hunk 11]
```diff
@@ -93,6 +97,20 @@ The basic operation of the tool is this:
...
+## Basic Operation for legacy version of ilib
+1. Write the `inc` file to specify which files to assemble.
+   If not, the default file that the tool (`./src/ilib-ilib-assemble-inc.js`)has will be used.
+2. Read the javscript files from the inc file list and extract the list of dependent JS and locale data.
+3. After the tool has been executed, the two style files will be placed the `output-dir`.
+   One is for the result of all JS files merged into a single usable file. If the file name is not specified, the default file name is `ilib-assemble.js`. and  another file is for the locale data per `[locale].js`.  
+   The data is written as follows: `ilib.data_[feature]_locale = {}`  
+   Here is an example of localeinfo for the ko-KR locale.
+   ```js
+   ilib.data.localeinfo_ko = {}
```
Comment (ehoogerbeets) [inline README.md]: Same with these

File: README.md [hunk 12]
```diff
@@ -93,6 +97,20 @@ The basic operation of the tool is this:
 The data can then be loaded by the ilib-localedata package for use in the various ilib-*
 packages.
 
+## Basic Operation for legacy version of ilib
+1. Write the `inc` file to specify which files to assemble.
+   If not, the default file that the tool (`./src/ilib-ilib-assemble-inc.js`)has will be used.
+2. Read the javscript files from the inc file list and extract the list of dependent JS and locale data.
+3. After the tool has been executed, the two style files will be placed the `output-dir`.
+   One is for the result of all JS files merged into a single usable file. If the file name is not specified, the default file name is `ilib-assemble.js`. and  another file is for the locale data per `[locale].js`.
```
Comment (ehoogerbeets) [inline README.md]: If you change the default file name below, then remember to fix this documentation as well

File: README.md (hunk 7 above)
Comment (gouniLee) [reply README.md:229]: I updated it.

File: src/index.js (hunk 1 above)
Comment (gouniLee) [reply src/index.js]: I updated it

File: src/index.js (hunk 2 above)
Comment (gouniLee) [reply src/index.js]: I updated it

File: src/index.js (hunk 3 above)
Comment (gouniLee) [reply src/index.js]: I updated it

File: src/index.js (hunk 4 above)
Comment (gouniLee) [reply src/index.js]: I updated it

File: src/legacyilibassemble.js (hunk 5 above)
Comment (gouniLee) [reply src/legacyilibassemble.js]: I update to have the different default name `ilib-all.js` and `ilib-all-inc.js`. so I also modified all the places where the related file names are written.

File: package.json (hunk 6 above)
Comment (gouniLee) [reply package.json]: You are right. It should be in dependencies. so I update it.

File: README.md (hunk 8 above)
Comment (gouniLee) [reply README.md]: I updated it

File: README.md (hunk 9 above)
Comment (gouniLee) [reply README.md]: oh! ^^;; I updated it. Thank you!

File: README.md (hunk 10 above)
Comment (gouniLee) [reply README.md]: I updated it

File: README.md (hunk 11 above)
Comment (gouniLee) [reply README.md]: I updated it

File: README.md (hunk 12 above)
Comment (gouniLee) [reply README.md]: yep. I also updated it

**Review by gouniLee — COMMENTED** [review]
I updated them following your comments. Please check them again. Thank you!

File: src/legacyilibassemble.js [hunk 13]
```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env node
```
Comment (ehoogerbeets) [inline src/legacyilibassemble.js]: One more thing that I missed earlier: you don't need this line if you are not executing this file as a separate command-line script. Please remove.

**Review by ehoogerbeets — APPROVED** [review]
Remove that one line and this is approved and can be merged

File: src/legacyilibassemble.js (hunk 13 above)
Comment (gouniLee) [reply src/legacyilibassemble.js]: Oh yes, It accidentally added because the file is generate by copying index.js at first time. Thank you.

**Pull request merged**